components/calendar/src/provider/hijri.rs

```diff
@@ -54,109 +47,119 @@ icu_provider::data_struct!(
-pub struct PackedHijriYearInfo(pub u8, pub u8);
+pub struct PackedHijriYearInfo(pub u16);
 
 impl PackedHijriYearInfo {
-    #[cfg(feature = "datagen")]
-    pub(crate) fn new(month_lengths: [bool; 12], ny_offset: i64) -> Self {
+    pub(crate) const fn new(
+        extended_year: i32,
+        month_lengths: [bool; 12],
+        start_day: RataDie,
+    ) -> Self {
+        let start_offset = start_day.const_diff(Self::mean_synodic_start_day(extended_year));
+
         debug_assert!(
-            -8 < ny_offset && ny_offset < 8,
+            -8 < start_offset && start_offset < 8,
             "Year offset too big to store"
         );
-        let ny_offset = ny_offset as i8;
+        let start_offset = start_offset as i8;
 
         let mut all = 0u16; // last byte unused
 
-        for (month, length_30) in month_lengths.iter().enumerate() {
-            if *length_30 {
-                all |= 1 << month as u16;
+        let mut i = 0;
+        while i < 12 {
+            #[allow(clippy::indexing_slicing)]
+            if month_lengths[i] {
+                all |= 1 << i;
             }
+            i += 1;
         }
 
-        if ny_offset < 0 {
+        if start_offset < 0 {
             all |= 1 << 12;
         }
-        all |= u16::from(ny_offset.unsigned_abs()) << 13;
-        let le = all.to_le_bytes();
-        Self(le[0], le[1])
+        all |= (start_offset.unsigned_abs() as u16) << 13;
+        Self(all)
     }
 
-    pub(crate) fn unpack(self) -> ([bool; 12], i64) {
-        let months = u16::from_le_bytes([self.0, self.1]);
+    pub(crate) fn unpack(self, extended_year: i32) -> ([bool; 12], RataDie) {
+        let month_lengths = core::array::from_fn(|i| self.0 & (1 << (i as u8) as u16) != 0);
+        let start_offset = if (self.0 & 0b1_0000_0000_0000) != 0 {
+            -((self.0 >> 13) as i64)
+        } else {
+            (self.0 >> 13) as i64
+        };
         (
-            core::array::from_fn(|i| months & (1 << (i as u8) as u16) != 0),
-            if (self.1 & 0b10000) != 0 {
-                -((self.1 >> 5) as i64)
-            } else {
-                (self.1 >> 5) as i64
-            },
+            month_lengths,
+            Self::mean_synodic_start_day(extended_year) + start_offset,
+        )
+    }
+
+    const fn mean_synodic_start_day(extended_year: i32) -> RataDie {
+        // -1 because the epoch is new year of year 1
+        // truncating instead of flooring does not matter, as this is used for positive years only
+        calendrical_calculations::islamic::ISLAMIC_EPOCH_FRIDAY.const_add(
+            ((extended_year - 1) as f64 * calendrical_calculations::islamic::MEAN_YEAR_LENGTH)
+                as i64,
         )
     }
 }
 
 impl AsULE for PackedHijriYearInfo {
-    type ULE = Self;
-    fn to_unaligned(self) -> Self {
-        self
+    type ULE = <u16 as AsULE>::ULE;
+    fn from_unaligned(unaligned: Self::ULE) -> Self {
+        Self(<u16 as AsULE>::from_unaligned(unaligned))
     }
```

Comment:
Issue: Why did you make this not transparent ULE any more? Aren't there invariants to be upheld?

Reply:
Because it's conceptually a `u16`, and only needs to be split into two `u8`s for ULE purposes.

Reply:
So, there are no invariants? Any `u16` is a valid `PackedHijriYearInfo`?

Reply:
Yep

Reply:
I didn't change the packing scheme, before it was `pub u8, pub u8`, where all patterns where valid, now it's `pub u16` with also all patterns valid. All I did was separate the packing logic from the ULE logic.

Reply:
ok, convinced given that it used to be `pub u8, pub u8`

(@Manishearth is this really the intent?)

Reply:
it uses 12 bits for the 12 months, and 4 bits for the compressed RD, so yes it seems very intentional (and efficient) to allow all bit patterns.

Reply:
Do we need all 4 bits for the RD modifier?

Reply:
Also, there's an implicit invariant that there are only ~half of the months that are 30-day months.

Reply:
We *allow* using four bits. For the concrete UAQ data the offset is only ever +/-1, but for the observational cache it might be bigger (hard to tell, the JSON data is not readable).

Reply:
You're welcome to fix this code, my change here does not affect any invariants or behaviour, it's all refactoring.